aarch64/kernel/memory/src/paging/mapper.rs

```diff
@@ -0,0 +1,1280 @@
+    pub fn translate_page(&self, page: Page) -> Option<Frame> {
+        let p3 = self.p4(true).next_table(page.p4_index(), true);
+
+        // Temporarily removing these checks while we
+        // Learn how to create huge pages on aarch64.
+        /*
+        let huge_page = || {
+            p3.and_then(|p3| {
+                let p3_entry = &p3[page.p3_index()];
+                // 1GiB page?
+                if let Some(start_frame) = p3_entry.pointed_frame() {
+                    if p3_entry.flags().is_huge() {
+                        // address must be 1GiB aligned
+                        assert!(start_frame.number() % (ENTRIES_PER_PAGE_TABLE * ENTRIES_PER_PAGE_TABLE) == 0);
+                        return Some(Frame::containing_address(PhysicalAddress::new_canonical(
+                            PAGE_SIZE * (start_frame.number() + page.p2_index() * ENTRIES_PER_PAGE_TABLE + page.p1_index())
+                        )));
+                    }
+                }
+                if let Some(p2) = p3.next_table(page.p3_index(), true) {
+                    let p2_entry = &p2[page.p2_index()];
+                    // 2MiB page?
+                    if let Some(start_frame) = p2_entry.pointed_frame() {
+                        if p2_entry.flags().is_huge() {
+                            // address must be 2MiB aligned
+                            assert!(start_frame.number() % ENTRIES_PER_PAGE_TABLE == 0);
+                            return Some(Frame::containing_address(PhysicalAddress::new_canonical(
+                                PAGE_SIZE * (start_frame.number() + page.p1_index())
+                            )));
+                        }
+                    }
+                }
+                None
+            })
+        };
+        */
+
+        p3.and_then(|p3| p3.next_table(page.p3_index(), true))
+            .and_then(|p2| p2.next_table(page.p2_index(), true))
+            .and_then(|p1| p1[page.p1_index()].pointed_frame())
+            // .or_else(huge_page)
+    }
+
+
+    /// An internal function that performs the actual mapping of a range of allocated `pages`
+    /// to a range of allocated `frames`.
+    /// 
+    /// Returns a tuple of the new `MappedPages` object containing the allocated `pages`
+    /// and the allocated `frames` object.
+    /// 
+    /// With `active` set to true, you indicate that the
+    /// modified table is currently used by the CPU, in which case
+    /// it will be accessed using the recursive paging special entry.
+    /// Otherwise, the code assumes we are in identity-mapping, and
+    /// the physical addresses in a table are used as virtual addresses.
+    pub(super) fn internal_map_to(
+        &mut self,
+        pages: AllocatedPages,
+        frames: AllocatedFrames,
+        flags: PteFlags,
+        active: bool,
+    ) -> Result<(MappedPages, AllocatedFrames), &'static str> {
+        // On aarch64, P4/P3/P2 should only contain valid
```

Comment:
some new changes in the main branch should address this. I've added a function called `PteFlagsArch::adjust_for_higher_level_pte()` that is the proper place to set these `higher_level_flags` (previously called `top_level_flags`)

Reply:
I changed the `aarch64` version of `adjust_for_higher_level_pte` so that it works; my implementation discards any flag set by the caller, which isn't great